### ISSUE TITLE
AUTOINCREMENT not generated on primary key with ValueConverter

### ISSUE THREAD

**Issue by pendenaor** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1117]
### Steps to reproduce
Create a console NET Core 3.1 project.
**TestEntityCustom.cs**
```cs
namespace TestEFCore
{
    public readonly struct TestId
    {
        private readonly uint id;

        private TestId(uint id)
        {
            this.id = id;
        }

        public uint ToId() => this.id;
        public static TestId From(uint id) => new TestId(id);
    }

    public class TestEntityCustom
    {
        public TestId Id { get; set; }
    }
}
```

**TestDbContext.cs**
```cs

using System;

using Microsoft.EntityFrameworkCore;

using Pomelo.EntityFrameworkCore.MySql.Infrastructure;

namespace TestEFCore
{
    public class TestDbContext : DbContext
    {
        public DbSet<TestEntityCustom> TestCustomEntities { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder
                // replace with your connection string
                .UseMySql(
                    "server=localhost;port=33060;user=xxxxx;password=yyyy;database=dbtest;charset=latin1",
                    mySqlOptions => mySqlOptions
                    // replace with your Server Version and Type
                        .ServerVersion(new Version(8, 0, 18), ServerType.MySql));
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<TestEntityCustom>()
                .Property(e => e.Id)
                .HasConversion(e => e.ToId(), e => TestId.From(e));
        }
    }
}
```

**program.cs**
```cs
using System;

using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;

using Pomelo.EntityFrameworkCore.MySql.Infrastructure;

namespace TestEFCore
{
    class Program
    {
        static void Main()
        {
        }
    }
}
```

### The issue
The AUTOINCREMENT statement is not generated in the SQL script when running `dotnet ef migrations script`. 
```sql
CREATE TABLE `TestCustomEntities` (
    `Id` int unsigned NOT NULL,
    CONSTRAINT `PK_TestCustomEntities` PRIMARY KEY (`Id`)
);
```

I've added `HasKey` et `ValueGeneratedOnAdd` when creating the model with no result.
If i replace `TestId` by `uint`, it work flawlessly.

Maybe related to #720 

### Further technical details

MySQL version: 5.7.28
Operating system: ubuntu 18.04 (server) / Win10 (client)
Pomelo.EntityFrameworkCore.MySql version: 3.1.1
Microsoft.AspNetCore.App version: 3.1.5

**Comment by pendenaor** [comment]
To understand this issue, I would like to debug `ColumnDefinition` method in `MySqlMigrationsSqlGenerator.cs`, how do i do that ?

**Comment by mguinness** [comment]
An unrelated observation - you state MySQL version is 5.7.28, but you use `ServerVersion(new Version(8, 0, 18), ServerType.MySql)`

**Comment by pendenaor** [comment]
Thanks @lauxjpn!
Digging into your PR, it appears that, for now, there is not workaround that i can put in my code. My only hope is waiting for version 3.2.0, isn't it?

**Comment by pendenaor** [comment]
> An unrelated observation - you state MySQL version is 5.7.28, but you use `ServerVersion(new Version(8, 0, 18), ServerType.MySql)`

yes, dumb copy-paste from documentation 😄

**Comment by lauxjpn** [comment]
> To understand this issue, I would like to debug ColumnDefinition method in MySqlMigrationsSqlGenerator.cs, how do i do that ?

The fix is already in #1118, but one way of debugging migration related code, is to insert `Debugger.Launch()` at the code position of interest.

After the PR is merged, the following will work and generate an `AUTO_INCREMENT` column (this will also work with migrations; it is not necessary for the type to use generics, I just defined it in this way for testing purposes):

```c#
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace IssueConsoleTemplate
{
    public readonly struct TestId<T>
    {
        private readonly T _id;

        private TestId(T id)
        {
            _id = id;
        }

        public T ToId() => _id;
        public static TestId<T> From(T id) => new TestId<T>(id);
    }

    public class TestEntityCustom
    {
        public TestId<uint> Id { get; set; }
    }
    
    public class Context : DbContext
    {
        public DbSet<TestEntityCustom> TestCustomEntities { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder
                .UseMySql("server=127.0.0.1;port=3306;user=root;password=;database=Issue1117", b => b.ServerVersion("8.0.20-mysql"))
                //.UseSqlServer(@"Data Source=.\MSSQL14;Integrated Security=SSPI;Initial Catalog=Issue1117")
                .UseLoggerFactory(
                    LoggerFactory.Create(
                        b => b
                            .AddConsole()
                            .AddFilter(level => level >= LogLevel.Information)))
                .EnableSensitiveDataLogging()
                .EnableDetailedErrors();
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<TestEntityCustom>()
                .Property(e => e.Id)
                .ValueGeneratedOnAdd()
                .HasConversion(e => e.ToId(), e => TestId<uint>.From(e));
        }
    }

    internal static class Program
    {
        private static void Main()
        {
            using var context = new Context();

            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();
        }
    }
}
```

It generates the following `CREATE TABLE` statement:

```sql
CREATE TABLE `TestCustomEntities` (
    `Id` int unsigned NOT NULL AUTO_INCREMENT,
    CONSTRAINT `PK_TestCustomEntities` PRIMARY KEY (`Id`)
);
```

> Digging into your PR, it appears that, for now, there is not workaround that i can put in my code. My only hope is waiting for version 3.2.0, isn't it?

As a **workaround**, you can use the following code:

```c#
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    modelBuilder.Entity<TestEntityCustom>()
        .Property(e => e.Id)
        //.ValueGeneratedOnAdd() // <-- works after PR has been merged
        .HasAnnotation("MySql:ValueGenerationStrategy", MySqlValueGenerationStrategy.IdentityColumn) // <-- works now
        .HasConversion(e => e.ToId(), e => TestId<uint>.From(e));
}
```

After the PR has been merged, you can use the latest [nightly build](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#nightly-builds).

**Comment by pendenaor** [comment]
Thanks again, @lauxjpn 👍
